### PR TITLE
SliceViewer: Exception when changing peak workspace in the ADS when overlaid

### LINKED ISSUE #30308: Sliceviewer exception when changing peaks workspace when overlaid
### Expected behavior
Having overlaid a peaks workspace in sliceviewer, if the peaks workspace were to change (e.g. integrated again with different radius) then the table and plot in the sliceviewer would be updated. Or at least the peaks workspace removed from sliceviewer.

### Actual behavior
The table and plot are not updated and when the user clicks on a peak in the table there is an unexpected exception
```
Traceback (most recent call last):
  File "C:\MantidNightlyInstall\bin\mantidqt\widgets\sliceviewer\peaksviewer\view.py", line 166, in _on_row_clicked
    self._row_selected()
  File "C:\MantidNightlyInstall\bin\mantidqt\widgets\sliceviewer\peaksviewer\view.py", line 172, in _row_selected
    self._presenter.notify(self._presenter.Event.PeakSelected)
  File "C:\MantidNightlyInstall\bin\mantidqt\widgets\sliceviewer\peaksviewer\presenter.py", line 89, in notify
    self._peak_selected()
  File "C:\MantidNightlyInstall\bin\mantidqt\widgets\sliceviewer\peaksviewer\presenter.py", line 123, in _peak_selected
    self._view.set_slicepoint(self.model.slicepoint(selected_index, self._view.sliceinfo))
  File "C:\MantidNightlyInstall\bin\mantidqt\widgets\sliceviewer\peaksviewer\model.py", line 90, in slicepoint
    peak = self.ws.getPeak(selected_index)
RuntimeError: Variable invalidated, data has been deleted.
```
### Steps to reproduce the behavior
For example
(1) Load in MD data and peak workspace
(2) Overlay peaks workspace
(3) Integrate the workspace using IntegratePeaksMD

### Platforms affected
found on unscripted testing of  workbench nightly install from 07/01/21

### PR DESCRIPTION
**Pull request by Pasarus**
**Description of work.**
Adds handles for the ADS observer in the peaksviewer presenter, which handles if a workspace is removed from underneath the sliceviewer if a peaks workspace is loaded in.

**To test:**
- Run this Script in workbench:
```
from mantid.simpleapi import *

SXD23767 = Load(Filename='SXD23767.raw', LoadMonitors='Exclude')

SetUB(SXD23767, 1,1,2,90,90,120)
md_non_ortho = ConvertToDiffractionMDWorkspace(InputWorkspace='SXD23767', OutputDimensions='HKL')
```
- Follow issue instructions to create peaks workspace
- Open slice viewer on md_non_ortho
- Add peaks workspace in using this button:
image
- Ensure the peaks workspace is displayed on the data display by double clicking the workspace table display on the right of the slice viewer
- Then remove the peaks workspace from the ADS, the UI should update to reflect the new ADS state.
- Repeat but for renaming the peaks workspace, replacing the peaks workspace, and clearing the ADS as a whole.
 

Fixes #30308 

---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

**Comment by RichardWaiteSTFC** [comment]
Nice work adding an observer but I have come accross a couple of problems:
(1) I think perhaps the ADS subscriber isn't deleted/unsubscribed when sliceviewer is closed. For example if you view some peaks and then close sliceviewer and try to delete the peaks workspace from the ADS then you get the following error
```
Error occurred in handler:
Traceback (most recent call last):
  File "C:\mantid\qt\python\mantidqt\widgets\sliceviewer\adsobsever.py", line 22, in wrapper
    func(*args, **kwargs)
  File "C:\mantid\qt\python\mantidqt\widgets\sliceviewer\adsobsever.py", line 60, in deleteHandle
    self.on_delete_workspace(ws_name)
  File "C:\mantid\qt\python\mantidqt\utils\qt\qappthreadcall.py", line 52, in __call__
    raise self._exc_info[1].with_traceback(self._exc_info[2])
  File "C:\mantid\qt\python\mantidqt\utils\qt\qappthreadcall.py", line 62, in on_call
    self.callee(*self._args, **self._kwargs)
  File "C:\mantid\qt\python\mantidqt\widgets\sliceviewer\peaksviewer\presenter.py", line 270, in delete_handle
    self.remove_peaksworkspace(ws_name)
  File "C:\mantid\qt\python\mantidqt\widgets\sliceviewer\peaksviewer\presenter.py", line 223, in remove_peaksworkspace
    self._view.remove_peaksviewer(child.view)
  File "C:\mantid\qt\python\mantidqt\widgets\sliceviewer\peaksviewer\view.py", line 212, in remove_peaksviewer
    item = layout.takeAt(layout.indexOf(widget))
RuntimeError: wrapped C/C++ object of type QVBoxLayout has been deleted
```
(2) Also one minor thing is that when you rename a workspace (or change it by e.g. integrating) it correctly reloads the new peaks workspace, however it does not preserve the order of the peaks workspaces in the RHS of the viewer (e.g. it will always add the new workspace to the bottom). This is not a big deal though, I don't think it's worth addressing in this PR.
(3) Very Very minor issue is that the peaks view button when selected does not stay depressed like e.g. the grid button. This is not to do with this PR though (it's on master and is an unrelated issue), just thought I'd mention it in case you could fix it quickly at the same time.

**Comment by Pasarus** [comment]
@RichardWaiteSTFC I have fixed issue 1
2. I agree, it effectively replaces it to update the reference for now unfortunately.
3. Not really to do with this PR and not quite a simple fix, I am not sure, will leave it for.

File: qt/python/mantidqt/widgets/sliceviewer/peaksviewer/test/test_peaksviewer_peaksviewercollectionpresenter.py [hunk 1]
```diff
@@ -82,6 +85,99 @@ def test_notify_called_for_each_subpresenter(self, mock_create_model):
...
+        presenter.remove_peaksworkspace = MagicMock()
+        presenter.workspace_names = MagicMock(return_value=["ws"])
+        presenter.append_peaksworkspace = MagicMock()
+
+        presenter.rename_handle("ws", "ws1")
+
+        presenter.remove_peaksworkspace.assert_called_once_with("ws")
+        presenter.append_peaksworkspace.assert_called_once_with("ws1")
+
+    def test_ensure_close_called
```
Comment (RichardWaiteSTFC) [inline qt/python/mantidqt/widgets/sliceviewer/peaksviewer/test/test_peaksviewer_peaksviewercollectionpresenter.py:179]: Did you intend to add another test here, or it it covered by the sliceviewer presenter tests?

**Review by RichardWaiteSTFC — CHANGES_REQUESTED** [review]

File: qt/python/mantidqt/widgets/sliceviewer/peaksviewer/test/test_peaksviewer_peaksviewercollectionpresenter.py (hunk 1 above)
Comment (Pasarus) [reply qt/python/mantidqt/widgets/sliceviewer/peaksviewer/test/test_peaksviewer_peaksviewercollectionpresenter.py:179]: AH I must have not finished that test! I will go do that now! This is why we have code review!

**Comment by RichardWaiteSTFC** [comment]
There seems to be a regression, if you open sliceviewer and then clear the ADS you get the following error
```
Error occurred in handler:
Traceback (most recent call last):
  File "C:\mantid\qt\python\mantidqt\widgets\sliceviewer\adsobsever.py", line 22, in wrapper
    func(*args, **kwargs)
  File "C:\mantid\qt\python\mantidqt\widgets\sliceviewer\adsobsever.py", line 55, in clearHandle
    self.on_clear()
  File "C:\mantid\qt\python\mantidqt\utils\qt\qappthreadcall.py", line 46, in __call__
    return self.callee(*args, **kwargs)
  File "C:\mantid\qt\python\mantidqt\widgets\sliceviewer\presenter.py", line 395, in ADS_cleared
    self.view.emit_close()
  File "C:\mantid\qt\python\mantidqt\widgets\observers\observing_view.py", line 41, in emit_close
    self.close_signal.emit()
RuntimeError: wrapped C/C++ object of type SliceViewerView has been deleted
Error occurred in handler:
Traceback (most recent call last):
  File "C:\mantid\qt\python\mantidqt\widgets\sliceviewer\adsobsever.py", line 22, in wrapper
    func(*args, **kwargs)
  File "C:\mantid\qt\python\mantidqt\widgets\sliceviewer\adsobsever.py", line 55, in clearHandle
    self.on_clear()
  File "C:\mantid\qt\python\mantidqt\utils\qt\qappthreadcall.py", line 46, in __call__
    return self.callee(*args, **kwargs)
  File "C:\mantid\qt\python\mantidqt\widgets\sliceviewer\presenter.py", line 395, in ADS_cleared
    self.view.emit_close()
  File "C:\mantid\qt\python\mantidqt\widgets\observers\observing_view.py", line 41, in emit_close
    self.close_signal.emit()
RuntimeError: wrapped C/C++ object of type SliceViewerView has been deleted
Error occurred in handler:
Traceback (most recent call last):
  File "C:\mantid\qt\python\mantidqt\widgets\sliceviewer\adsobsever.py", line 22, in wrapper
    func(*args, **kwargs)
  File "C:\mantid\qt\python\mantidqt\widgets\sliceviewer\adsobsever.py", line 55, in clearHandle
    self.on_clear()
  File "C:\mantid\qt\python\mantidqt\utils\qt\qappthreadcall.py", line 46, in __call__
    return self.callee(*args, **kwargs)
  File "C:\mantid\qt\python\mantidqt\widgets\sliceviewer\presenter.py", line 395, in ADS_cleared
    self.view.emit_close()
  File "C:\mantid\qt\python\mantidqt\widgets\observers\observing_view.py", line 41, in emit_close
    self.close_signal.emit()
RuntimeError: wrapped C/C++ object of type SliceViewerView has been deleted
Error occurred in handler:
Traceback (most recent call last):
  File "C:\mantid\qt\python\mantidqt\widgets\sliceviewer\adsobsever.py", line 22, in wrapper
    func(*args, **kwargs)
  File "C:\mantid\qt\python\mantidqt\widgets\sliceviewer\adsobsever.py", line 55, in clearHandle
    self.on_clear()
  File "C:\mantid\qt\python\mantidqt\utils\qt\qappthreadcall.py", line 46, in __call__
    return self.callee(*args, **kwargs)
  File "C:\mantid\qt\python\mantidqt\widgets\sliceviewer\presenter.py", line 395, in ADS_cleared
    self.view.emit_close()
  File "C:\mantid\qt\python\mantidqt\widgets\observers\observing_view.py", line 41, in emit_close
    self.close_signal.emit()
RuntimeError: wrapped C/C++ object of type SliceViewerView has been deleted
```

It seems to try and close the view twice?

**Comment by RichardWaiteSTFC** [comment]
Looks like it's failing a test, other than that it seems to work very well and the ADS observer now behaves. There is one slight bug in that if you add a peak to a peak table that is overlaid by e.g. 
```
p = mtd['peaks_workspace']
pk = p.createPeak([2,2,2])
p.addPeak(pk)
```
the table in the sliceviewer is not updated however no exception is caused when navigating to peaks, and the workspace can be taken off and overlaid again (no problem exists for deleting peaks, the table is updated).

**Review by RichardWaiteSTFC — APPROVED** [review]

**Pull request merged**